File: src/main/java/aurocosh/divinefavor/common/receipes/RecipeBuilder.java
package aurocosh.divinefavor.common.receipes;

import aurocosh.divinefavor.common.item.calling_stones.ItemCallingStone;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.Ingredient;

import java.util.ArrayList;
import java.util.List;

public class RecipeBuilder {
    public ItemStack result;
    private Ingredient callingStone;
    private List<Ingredient> ingredients;

    public RecipeBuilder(ItemStack result, ItemCallingStone callingStone) {
        this.result = result;
        this.callingStone = Ingredient.fromStacks(new ItemStack(callingStone));
        ingredients = new ArrayList<>();
    }

    public RecipeBuilder(ItemStack result, ItemCallingStone callingStone, List<ItemStack> ingredientStacks) {
        this(result, callingStone);
        for (ItemStack stack : ingredientStacks)
            ingredients.add(Ingredient.fromStacks(stack));
    }

    public RecipeBuilder addIngredient(Item item, int count) {
        ingredients.add(Ingredient.fromStacks(new ItemStack(item, count)));
        return this;
    }

    public RecipeBuilder addIngredient(Item item) {
        ingredients.add(Ingredient.fromItems(item));
        return this;
    }

    public ImmaterialMediumRecipe create() {
        Ingredient[] ingredientsArray = new Ingredient[ingredients.size()];
        ingredientsArray = ingredients.toArray(ingredientsArray);

        return new ImmaterialMediumRecipe(result, callingStone, ingredientsArray);
    }
}

File: src/main/kotlin/aurocosh/divinefavor/common/receipes/RecipeBuilder.kt
package aurocosh.divinefavor.common.receipes

import aurocosh.divinefavor.common.item.gems.calling_stones.ItemCallingStone
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.Ingredient
import java.util.*

class RecipeBuilder(var result: ItemStack, callingStone: ItemCallingStone) {
    private val callingStone: Ingredient = Ingredient.fromStacks(ItemStack(callingStone))
    private val ingredients: MutableList<Ingredient> = ArrayList()

    constructor(result: ItemStack, callingStone: ItemCallingStone, ingredientStacks: List<ItemStack>) : this(result, callingStone) {
        for (stack in ingredientStacks)
            ingredients.add(Ingredient.fromStacks(stack))
    }

    fun addIngredient(item: Item, count: Int): RecipeBuilder {
        ingredients.add(Ingredient.fromStacks(ItemStack(item, count)))
        return this
    }

    fun addIngredient(item: Item): RecipeBuilder {
        ingredients.add(Ingredient.fromItems(item))
        return this
    }

    fun create(): ImmaterialMediumRecipe {
        return ImmaterialMediumRecipe(result, callingStone, ingredients)
    }
}
